### ISSUE TITLE
Proposal: unify kriging routines

### ISSUE THREAD

**Issue by MuellerSeb** [GeoStat-Framework/GSTools#79]
Since all kriging routines could be reformulated in order to use the covariance matrix instead of the variogram matrix, we could unify the kriging base class to serve as a swiss knife for kriging.

All other kriging routines are then just a shortcut with a limited input parameter list.

**Comment by MuellerSeb** [comment]
It would also be good to use the pseudo-inverse, if the kriging matrix is singular.
We could check this as suggested here:
https://stackoverflow.com/a/9155489/6696397

When using the Covariance matrix, we could also use SVD to invert the matrix as suggested here:
https://www.quora.com/How-do-I-get-the-inverse-of-a-matrix-using-SVD-in-Python/answer/Samuel-S-Watson-1

The full inverted kriging matrix could then be calculated by using the inversion rules for block matrices:
https://en.wikipedia.org/wiki/Block_matrix#Block_matrix_inversion

**Comment by LSchueler** [comment]
Great idea! And nice research, the solution from the stackoverflow link is pretty funny.

What about performance? - Do you have any idea at what expense the flexibility will come?

**Comment by MuellerSeb** [comment]
The try-except block doesn't include any overhead, when the current implementation succeeds. The pinv takes about 5 times more time to compute the inverse (if we have numpy 1.17, which brings the `hermitian` keyword):
```python
import numpy as np

a = np.random.random((1000, 1000))
a += a.T

import timeit

%timeit np.linalg.inv(a)
94.8 ms ± 5.18 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

%timeit np.linalg.pinv(a)
1.17 s ± 76.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

%timeit np.linalg.pinv(a, hermitian=True)
452 ms ± 25.6 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

**Comment by LSchueler** [comment]
Actually, I mean the fully flexible approach, not just the pseudo inverse. That's a thing we should definitely do.

**Comment by MuellerSeb** [comment]
Sorry for getting you wrong. The flexible approach is not including any overhead. At the time of coding I was just not aware, that kriging can always be formulated using the covariance function instead of the variogram function (in case of finite sill and stationary models, which we always assume). So we get rid of the different approaches for creating the kriging matrix, since every method only extends the matrix:

simple -> ordinary -> (ext)drift

And I propose to simply use this hierarchy and extend the matrix if needed.

**Comment by MuellerSeb** [comment]
As seen in #89, sometimes a `LinAlgError` isn't even detected, so we should stick to the `pinv`from the beginning.
@LSchueler, do you think it would be ok, to increase the minimal required version of numpy to `v1.17`?

**Comment by MuellerSeb** [comment]
I just noticed, that `scipy 1.1.0` already provides a pseudo inverse algorithm for symmetric matrices:
https://docs.scipy.org/doc/scipy-1.1.0/reference/generated/scipy.linalg.pinvh.html#scipy.linalg.pinvh

So we could use this one, without increasing any required versions.

**Comment by MuellerSeb** [comment]
This was resolved with #97
